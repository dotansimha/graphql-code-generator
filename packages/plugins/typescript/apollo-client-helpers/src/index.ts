import { Types, PluginValidateFn, PluginFunction } from '@graphql-codegen/plugin-helpers';
import { GraphQLSchema, isInterfaceType, isObjectType } from 'graphql';
import { extname } from 'path';
import { ApolloClientHelpersConfig } from './config';

export const plugin: PluginFunction<ApolloClientHelpersConfig> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: ApolloClientHelpersConfig
) => {
  const results: Types.ComplexPluginOutput[] = [];

  results.push(generateTypePoliciesSignature(schema, config));

  return {
    prepend: results.reduce((prev, r) => [...prev, ...(r.prepend || [])], []),
    append: results.reduce((prev, r) => [...prev, ...(r.append || [])], []),
    content: results.map(r => r.content).join('\n'),
  };
};

function generateTypePoliciesSignature(
  schema: GraphQLSchema,
  config: ApolloClientHelpersConfig
): Types.ComplexPluginOutput {
  const typeMap = schema.getTypeMap();
  const perTypePolicies: string[] = [];
  const typedTypePolicies = Object.keys(typeMap).reduce((prev, typeName) => {
    const type = typeMap[typeName];

    if (!typeName.startsWith('__') && (isObjectType(type) || isInterfaceType(type))) {
      const fieldsNames = Object.keys(type.getFields()).filter(f => !f.startsWith('__'));
      const keySpecifierVarName = `${typeName}KeySpecifier`;
      const fieldPolicyVarName = `${typeName}FieldPolicy`;

      perTypePolicies.push(
        `export type ${keySpecifierVarName} = (${fieldsNames
          .map(f => `'${f}'`)
          .join(' | ')} | ${keySpecifierVarName})[];`
      );

      perTypePolicies.push(`export type ${fieldPolicyVarName} = {
${fieldsNames.map(fieldName => `\t${fieldName}?: FieldPolicy<any> | FieldReadFunction<any>`).join(',\n')}
};`);

      return {
        ...prev,
        [typeName]: `Omit<TypePolicy, "fields" | "keyFields"> & {
\t\tkeyFields${
          config.requireKeyFields ? '' : '?'
        }: false | ${keySpecifierVarName} | (() => undefined | ${keySpecifierVarName}),
\t\tfields?: ${fieldPolicyVarName},
\t}`,
      };
    }

    return prev;
  }, {} as Record<string, string>);

  const rootTypes = [
    schema.getQueryType()?.name,
    schema.getMutationType()?.name,
    schema.getSubscriptionType()?.name,
  ].filter(Boolean);

  const rootContent = `export type TypedTypePolicies = TypePolicies & {${Object.keys(typedTypePolicies)
    .map(typeName => {
      const nonOptional = config.requirePoliciesForAllTypes && !rootTypes.includes(typeName);

      return `\n\t${typeName}${nonOptional ? '' : '?'}: ${typedTypePolicies[typeName]}`;
    })
    .join(',')}\n};`;

  return {
    prepend: [
      `import ${
        config.useTypeImports ? 'type ' : ''
      }{ FieldPolicy, FieldReadFunction, TypePolicies, TypePolicy } from '@apollo/client/cache';`,
    ],
    content: [...perTypePolicies, rootContent].join('\n'),
  };
}

export const validate: PluginValidateFn<ApolloClientHelpersConfig> = async (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config,
  outputFile: string
) => {
  if (extname(outputFile) !== '.ts' && extname(outputFile) !== '.tsx') {
    throw new Error(`Plugin "apollo-client-helpers" requires extension to be ".ts" or ".tsx"!`);
  }
};
