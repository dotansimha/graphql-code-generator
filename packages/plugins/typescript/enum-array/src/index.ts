import { GraphQLSchema, GraphQLEnumType } from 'graphql';
import { PluginFunction, Types } from '@graphql-codegen/plugin-helpers';
import { EnumArrayPluginConfig } from './config';

function getEnumTypeMap(schema: GraphQLSchema): GraphQLEnumType[] {
  const typeMap = schema.getTypeMap();
  const result: GraphQLEnumType[] = [];
  for (const key in typeMap) {
    if (typeMap[key].astNode?.kind === 'EnumTypeDefinition') {
      result.push(typeMap[key] as GraphQLEnumType);
    }
  }
  return result;
}

function buildArrayDefinition(e: GraphQLEnumType): string {
  const upperName = e.name
    .replace(/[A-Z]/g, letter => `_${letter}`)
    .slice(1)
    .toUpperCase();
  const values = e
    .getValues()
    .map(v => `'${v.value}'`)
    .join(', ');
  return `export const ${upperName}: ${e.name}[] = [${values}];`;
}

function buildImportStatement(enums: GraphQLEnumType[], importFrom: string): string[] {
  const names: string[] = Object.values(enums).map(e => e.name);
  return [`import { ${names.join(', ')} } from "${importFrom}";`];
}

export const plugin: PluginFunction<EnumArrayPluginConfig> = async (
  schema: GraphQLSchema,
  _documents: Types.DocumentFile[],
  config: EnumArrayPluginConfig
): Promise<Types.PluginOutput> => {
  const importFrom: EnumArrayPluginConfig['importFrom'] = config.importFrom;
  const enums = getEnumTypeMap(schema);
  const content = enums.map(buildArrayDefinition).join('\n');

  const result: Types.PluginOutput = { content };
  if (importFrom) {
    result['prepend'] = buildImportStatement(enums, importFrom);
  }
  return result;
};

export default { plugin };
