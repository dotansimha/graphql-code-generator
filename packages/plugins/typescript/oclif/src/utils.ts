import { NamedTypeNode, OperationDefinitionNode, TypeNode, VariableDefinitionNode } from 'graphql';

type OclifFlagType = 'boolean' | 'integer' | 'string' | 'enum';

interface TypeConfig {
  required: boolean;
  list: boolean;
  innerType: NamedTypeNode;
}

export const getFlagConfigForVariableDefinition = (definition: VariableDefinitionNode) => {
  const { list, required, innerType } = getInnerType(definition.type);
  const oclifType = mapVariableTypeToOclifType(innerType);
  const parser = getParserForType(innerType);

  return `${definition.variable.name.value}: flags.${oclifType}({
  multiple: ${list},
  required: ${required},${parser ? `\n  parse: ${parser}` : ''}
})`;
};

// Supply a custom parser for oclif flag configuration
const getParserForType = (type: NamedTypeNode): string | void => {
  if (type.name.value === 'Float') {
    return 'input => Number(input)';
  }
};

const mapVariableTypeToOclifType = (type: NamedTypeNode): OclifFlagType => {
  if (type.name.value === 'Boolean') {
    return 'boolean';
  } else if (['Float', 'Int'].includes(type.name.value)) {
    // A quirk of oclif is that "integer" allows for any `number`-typed response, and then
    //   we supply our own parsing function to make sure it's a float and not an integer
    return 'integer';
  } else {
    return 'string';
  }
};

// Retrieve the inner type if nested within List and/or NonNull
const getInnerType = (type: TypeNode): TypeConfig => {
  let result: Partial<TypeConfig> = {
    list: false,
    required: false,
  };
  let _type = type;
  while (_type.kind !== 'NamedType') {
    if (_type.kind === 'ListType') {
      result.list = true;
    } else if (_type.kind === 'NonNullType') {
      result.required = true;
    }
    _type = _type.type;
  }
  result.innerType = _type;
  return result as TypeConfig;
};

// remove all @oclif directives from the document for transmission to the server
export const omitOclifDirectives = (node: OperationDefinitionNode) => {
  const directives = node.directives.filter(directive => directive.name.value !== 'oclif');
  return Object.assign({}, node, { directives });
};
