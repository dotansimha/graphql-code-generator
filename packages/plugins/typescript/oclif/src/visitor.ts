import {
  ClientSideBaseVisitor,
  ClientSideBasePluginConfig,
  LoadedFragment,
  indentMultiline,
} from '@graphql-codegen/visitor-plugin-common';
import autoBind from 'auto-bind';
import { GraphQLSchema, OperationDefinitionNode, print } from 'graphql';

import { Config, Info } from '.';
import { getFlagConfigForVariableDefinition, omitOclifDirectives } from './utils';

interface Operation {
  node: OperationDefinitionNode;
  documentVariableName: string;
  operationType: string;
  operationResultType: string;
  operationVariablesTypes: string;
}

interface OclifDirectiveValues {
  description?: string;
  examples?: string[];
}

export class GraphQLRequestVisitor extends ClientSideBaseVisitor<Config, ClientSideBasePluginConfig> {
  private _operationsToInclude: Operation[] = [];
  private _info: Info;

  constructor(schema: GraphQLSchema, fragments: LoadedFragment[], rawConfig: Config, info: Info) {
    super(schema, fragments, rawConfig, {});
    this._info = info;

    const { handlerPath = '../../handler' } = rawConfig;

    // FIXME: This is taken in part from
    //  presets/near-operation-file/src/index.ts:139. How do I build a path relative to the outputFile in the same way?
    //  A plugin doesn't appear to have access to the same "options.baseOutputDir" that the preset does.
    // const absClientPath = resolve(info.outputFile, join(options.baseOutputDir, options.presetConfig.baseTypesPath));

    autoBind(this);
    this._additionalImports.push(`import { Command, flags } from '@oclif/command'`);
    this._additionalImports.push(`import handler from '${handlerPath}'`);
  }

  protected buildOperation(
    node: OperationDefinitionNode,
    documentVariableName: string,
    operationType: string,
    operationResultType: string,
    operationVariablesTypes: string
  ): string {
    this._operationsToInclude.push({
      node,
      documentVariableName,
      operationType,
      operationResultType,
      operationVariablesTypes,
    });

    return null;
  }

  // Clean client-side content (ie directives) out of the GraphQL document prior to sending to the server
  public get definition(): string {
    const operation = this._operationsToInclude[0];
    const clientOperation = print(omitOclifDirectives(operation.node));
    return `const ${operation.documentVariableName} = \`\n${clientOperation}\``;
  }

  // Generate the code required for this CLI operation
  public get cliContent(): string {
    if (this._operationsToInclude.length !== 1) {
      throw new Error(
        `Each graphql document should have exactly one operation; found ${this._operationsToInclude.length} while generating ${this._info.outputFile}.`
      );
    }

    const operation = this._operationsToInclude[0];

    // Find the @oclif directive in the client document, if it's there
    const directive = operation.node.directives.find(directive => directive.name.value === 'oclif');

    // Remap the directive's fields ie @oclif(description: "a name") to a more usable format
    const directiveValues: OclifDirectiveValues = {};

    if (directive) {
      directiveValues.examples = [];
      directive.arguments.forEach(arg => {
        const value = 'value' in arg.value ? arg.value.value.toString() : null;
        const { value: name } = arg.name;
        if (name === 'description') {
          directiveValues.description = value;
        } else if (name === 'example') {
          directiveValues.examples.push(value);
        } else {
          throw new Error(`Invalid field supplied to @oclif directive: ${name}`);
        }
      });
    }

    const { description, examples } = directiveValues;

    const flags = operation.node.variableDefinitions.map(getFlagConfigForVariableDefinition);

    return `
${this.definition}

export default class ${operation.node.name.value} extends Command {
  ${description ? `\nstatic description = "${description}";\n` : ''}
  ${examples ? `\nstatic examples: string[] = ${JSON.stringify(examples)};\n` : ''}
  static flags = {
    help: flags.help({ char: 'h' }),
${indentMultiline(flags.join(',\n'), 2)}
  };

  async run() {
    const { flags } = this.parse(${operation.node.name.value});
    await handler({ command: this, query: ${operation.documentVariableName}, variables: flags });
  }
}
`;
  }
}
