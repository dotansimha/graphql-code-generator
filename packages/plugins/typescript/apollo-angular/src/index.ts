import { Types, PluginValidateFn, PluginFunction } from '@graphql-codegen/plugin-helpers';
import { visit, GraphQLSchema, concatAST, Kind, FragmentDefinitionNode, OperationDefinitionNode } from 'graphql';
import { LoadedFragment } from '@graphql-codegen/visitor-plugin-common';
import { ApolloAngularVisitor } from './visitor';
import { extname } from 'path';
import gql from 'graphql-tag';
import { ApolloAngularRawPluginConfig } from './config';

export const plugin: PluginFunction<ApolloAngularRawPluginConfig> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config
) => {
  const allAst = concatAST(documents.map(v => v.document));
  const operations = allAst.definitions.filter(d => d.kind === Kind.OPERATION_DEFINITION) as OperationDefinitionNode[];
  const allFragments: LoadedFragment[] = [
    ...(allAst.definitions.filter(d => d.kind === Kind.FRAGMENT_DEFINITION) as FragmentDefinitionNode[]).map(
      fragmentDef => ({
        node: fragmentDef,
        name: fragmentDef.name.value,
        onType: fragmentDef.typeCondition.name.value,
        isExternal: false,
      })
    ),
    ...(config.externalFragments || []),
  ];

  const visitor = new ApolloAngularVisitor(schema, allFragments, operations, config, documents);
  const visitorResult = visit(allAst, { leave: visitor });

  return {
    prepend: visitor.getImports(),
    content: [
      visitor.fragments,
      ...visitorResult.definitions.filter(t => typeof t === 'string'),
      config.sdkClass ? visitor.sdkClass : null,
    ]
      .filter(a => a)
      .join('\n'),
  };
};

export const addToSchema = gql`
  directive @NgModule(module: String!) on OBJECT | FIELD
  directive @namedClient(name: String!) on OBJECT | FIELD
`;

export const validate: PluginValidateFn<any> = async (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config,
  outputFile: string
) => {
  if (extname(outputFile) !== '.ts') {
    throw new Error(`Plugin "apollo-angular" requires extension to be ".ts"!`);
  }
};

export { ApolloAngularVisitor };
