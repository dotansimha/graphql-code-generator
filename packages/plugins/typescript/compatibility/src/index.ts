import { GraphQLSchema, concatAST, visit } from 'graphql';
import { PluginFunction, Types } from '@graphql-codegen/plugin-helpers';
import { CompatibilityPluginVisitor } from './visitor';
import { CompatibilityPluginRawConfig } from './config';

const REACT_APOLLO_PLUGIN_NAME = 'typescript-react-apollo';

export const plugin: PluginFunction<CompatibilityPluginRawConfig> = async (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: CompatibilityPluginRawConfig,
  additionalData
): Promise<string> => {
  const allAst = concatAST(documents.map(v => v.document));

  const reactApollo = ((additionalData || {}).allPlugins || []).find(
    p => Object.keys(p)[0] === REACT_APOLLO_PLUGIN_NAME
  );
  const visitor = new CompatibilityPluginVisitor(config, schema, {
    reactApollo: reactApollo
      ? {
          ...(config || {}),
          ...(reactApollo[REACT_APOLLO_PLUGIN_NAME] as object),
        }
      : null,
  });

  const visitorResult = visit(allAst, {
    leave: visitor as any,
  });

  const discriminateUnion = `type DiscriminateUnion<T, U> = T extends U ? T : never;\n`;
  const result: string = visitorResult.definitions.filter(a => a && typeof a === 'string').join('\n');

  return result.includes('DiscriminateUnion') ? [discriminateUnion, result].join('\n') : result;
};
