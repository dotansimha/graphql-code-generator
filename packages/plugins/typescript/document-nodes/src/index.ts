import { PluginFunction, PluginValidateFn, Types } from '@graphql-codegen/plugin-helpers';
import {
  NamingConvention,
  LoadedFragment,
  RawClientSideBasePluginConfig,
} from '@graphql-codegen/visitor-plugin-common';
import { GraphQLSchema, visit, concatAST, FragmentDefinitionNode, Kind } from 'graphql';
import { TypeScriptDocumentNodesVisitor } from './visitor';

/**
 * @description This plugin generates TypeScript source (`.ts`) file from GraphQL files (`.graphql`).
 */
export interface TypeScriptDocumentNodesRawPluginConfig extends RawClientSideBasePluginConfig {
  /**
   * @default change-case-all#pascalCase
   * @description Allow you to override the naming convention of the output.
   * You can either override all namings, or specify an object with specific custom naming convention per output.
   * The format of the converter must be a valid `module#method`.
   * Allowed values for specific output are: `typeNames`, `enumValues`.
   * You can also use "keep" to keep all GraphQL names as-is.
   * Additionally you can set `transformUnderscore` to `true` if you want to override the default behavior,
   * which is to preserves underscores.
   *
   * Available case functions in `change-case-all` are `camelCase`, `capitalCase`, `constantCase`, `dotCase`, `headerCase`, `noCase`, `paramCase`, `pascalCase`, `pathCase`, `sentenceCase`, `snakeCase`, `lowerCase`, `localeLowerCase`, `lowerCaseFirst`, `spongeCase`, `titleCase`, `upperCase`, `localeUpperCase` and `upperCaseFirst`
   * [See more](https://github.com/btxtiger/change-case-all)
   *
   * @exampleMarkdown
   * ## Override All Names
   * ```yml
   * config:
   *   namingConvention: change-case-all#lowerCase
   * ```
   *
   * ## Upper-case enum values
   * ```yml
   * config:
   *   namingConvention:
   *     typeNames: change-case-all#pascalCase
   *     enumValues: change-case-all#upperCase
   * ```
   *
   * ## Keep name as-is
   * ```yml
   * config:
   *   namingConvention: keep
   * ```
   *
   * ## Remove Underscores
   * ```yml
   * config:
   *   namingConvention:
   *     typeNames: change-case-all#pascalCase
   *     transformUnderscore: true
   * ```
   */
  namingConvention?: NamingConvention;
  /**
   * @default ""
   * @description Adds prefix to the name
   *
   * @exampleMarkdown
   * ```yml
   *  documents: src/api/user-service/queries.graphql
   *  generates:
   *    src/api/user-service/queries.ts:
   *    plugins:
   *      - typescript-document-nodes
   *    config:
   *      namePrefix: 'gql'
   * ```
   */
  namePrefix?: string;
  /**
   * @default ""
   * @description Adds suffix to the name
   *
   * @exampleMarkdown
   * ```yml
   *  documents: src/api/user-service/queries.graphql
   *  generates:
   *    src/api/user-service/queries.ts:
   *    plugins:
   *      - typescript-document-nodes
   *    config:
   *      nameSuffix: 'Query'
   * ```
   */
  nameSuffix?: string;
  /**
   * @default ""
   * @description Adds prefix to the fragment variable
   */
  fragmentPrefix?: string;
  /**
   * @default ""
   * @description Adds suffix to the fragment variable
   */
  fragmentSuffix?: string;
}

export const plugin: PluginFunction<TypeScriptDocumentNodesRawPluginConfig> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: TypeScriptDocumentNodesRawPluginConfig
) => {
  const allAst = concatAST(documents.map(v => v.document));

  const allFragments: LoadedFragment[] = [
    ...(allAst.definitions.filter(d => d.kind === Kind.FRAGMENT_DEFINITION) as FragmentDefinitionNode[]).map(
      fragmentDef => ({
        node: fragmentDef,
        name: fragmentDef.name.value,
        onType: fragmentDef.typeCondition.name.value,
        isExternal: false,
      })
    ),
    ...(config.externalFragments || []),
  ];

  const visitor = new TypeScriptDocumentNodesVisitor(schema, allFragments, config, documents);
  const visitorResult = visit(allAst, { leave: visitor });

  return {
    prepend: visitor.getImports(),
    content: [visitor.fragments, ...visitorResult.definitions.filter(t => typeof t === 'string')].join('\n'),
  };
};

export const validate: PluginValidateFn<any> = async (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: any,
  outputFile: string
) => {
  if (!outputFile.endsWith('.ts')) {
    throw new Error(`Plugin "typescript-document-nodes" requires extension to be ".ts"!`);
  }
};
