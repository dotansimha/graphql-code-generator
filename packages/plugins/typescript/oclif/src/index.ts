import { Types, PluginValidateFn, PluginFunction, oldVisit } from '@graphql-codegen/plugin-helpers';
import { GraphQLSchema, concatAST, Kind, FragmentDefinitionNode } from 'graphql';
import { RawClientSideBasePluginConfig, LoadedFragment } from '@graphql-codegen/visitor-plugin-common';
import { GraphQLRequestVisitor } from './visitor';
import { extname } from 'path';

export interface Config extends RawClientSideBasePluginConfig {
  handlerPath?: string;
}

export interface Info {
  outputFile: string;
  allPlugins: any[];
}

export const plugin: PluginFunction = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: Config,
  info: Info
) => {
  const allAst = concatAST(
    documents.reduce((prev, v) => {
      return [...prev, v.document];
    }, [])
  );
  const allFragments: LoadedFragment[] = [
    ...(allAst.definitions.filter(d => d.kind === Kind.FRAGMENT_DEFINITION) as FragmentDefinitionNode[]).map(
      fragmentDef => ({
        node: fragmentDef,
        name: fragmentDef.name.value,
        onType: fragmentDef.typeCondition.name.value,
        isExternal: false,
      })
    ),
    ...(config.externalFragments || []),
  ];
  const visitor = new GraphQLRequestVisitor(schema, allFragments, config, info);
  oldVisit(allAst, { leave: visitor });

  return {
    prepend: visitor.getImports(),
    content: visitor.cliContent,
  };
};

export const validate: PluginValidateFn<any> = async (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: Config,
  outputFile: string
) => {
  if (extname(outputFile) !== '.ts') {
    throw new Error(`Plugin "typescript-oclif" requires output file extensions to be ".ts"!`);
  }
};

export { GraphQLRequestVisitor };
