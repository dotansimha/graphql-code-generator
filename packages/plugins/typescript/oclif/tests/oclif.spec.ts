import { plugin } from '../src/index';
import { parse, buildClientSchema } from 'graphql';
import { Types } from '@graphql-codegen/plugin-helpers';

const schema = buildClientSchema(require('../../../../../dev-test/githunt/schema.json'));

describe('oclif', () => {
  describe('cli', () => {
    it('builds from a read query with single string parameter', async () => {
      const document = parse(`
        query GetEntry($name: String!) {
          entry(repoFullName: $name) {
            repository {
              name
            }
          }
        }
      `);

      const config = {};
      const docs = [{ location: '', document }];
      const result = (await plugin(schema, docs, config, {
        outputFile: 'graphql.ts',
      })) as Types.ComplexPluginOutput;

      expect(result.content).toMatchSnapshot();
    });

    it('builds from a read query with single enum parameter', async () => {
      const document = parse(`
        query GetFeed($type: FeedType!) {
          feed(type: $type) {
            id
            commentCount
          }
        }
      `);

      const config = {};
      const docs = [{ location: '', document }];
      const result = (await plugin(schema, docs, config, {
        outputFile: 'graphql.ts',
      })) as Types.ComplexPluginOutput;

      expect(result.content).toMatchSnapshot();
    });

    it('builds from a read query with integer parameters', async () => {
      const document = parse(`
        query GetFeed($type: FeedType!, $offset: Int!, $limit: Int!) {
          feed(type: $type, offset: $offset, limit: $limit) {
            id
            commentCount
          }
        }
      `);

      const config = {};
      const docs = [{ location: '', document }];
      const result = (await plugin(schema, docs, config, {
        outputFile: 'graphql.ts',
      })) as Types.ComplexPluginOutput;

      expect(result.content).toMatchSnapshot();
    });

    it('reads metadata from the @oclif directive', async () => {
      const document = parse(`
        query GetCurrentUser @oclif(description: "Get Current User", example: "cli get-current-user", example: "hello") {
          currentUser {
            login
          }
        }
      `);

      const config = {};
      const docs = [{ location: '', document }];
      const result = (await plugin(schema, docs, config, {
        outputFile: 'graphql.ts',
      })) as Types.ComplexPluginOutput;

      expect(result.content).toMatchSnapshot();
    });

    it('builds from a mutation with string and enum type', async () => {
      const document = parse(`
        mutation Vote($name: String!, $type: VoteType!) {
          vote(repoFullName: $name, type: $type) {
            score
          }
        }
      `);

      const config = {};
      const docs = [{ location: '', document }];
      const result = (await plugin(schema, docs, config, {
        outputFile: 'graphql.ts',
      })) as Types.ComplexPluginOutput;

      expect(result.content).toMatchSnapshot();
    });
  });
});
