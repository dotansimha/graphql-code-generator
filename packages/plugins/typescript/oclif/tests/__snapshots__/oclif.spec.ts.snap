// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oclif cli builds from a mutation with string and enum type 1`] = `
"
const VoteDocument = \`
mutation Vote($name: String!, $type: VoteType!) {
  vote(repoFullName: $name, type: $type) {
    score
  }
}\`

export default class Vote extends Command {
  
  
  static flags = {
    help: flags.help({ char: 'h' }),
    name: flags.string({
      multiple: false,
      required: true,
    }),
    type: flags.string({
      multiple: false,
      required: true,
    })
  };

  async run() {
    const { flags } = this.parse(Vote);
    await handler({ command: this, query: VoteDocument, variables: flags });
  }
}
"
`;

exports[`oclif cli builds from a read query with integer parameters 1`] = `
"
const GetFeedDocument = \`
query GetFeed($type: FeedType!, $offset: Int!, $limit: Int!) {
  feed(type: $type, offset: $offset, limit: $limit) {
    id
    commentCount
  }
}\`

export default class GetFeed extends Command {
  
  
  static flags = {
    help: flags.help({ char: 'h' }),
    type: flags.string({
      multiple: false,
      required: true,
    }),
    offset: flags.integer({
      multiple: false,
      required: true,
    }),
    limit: flags.integer({
      multiple: false,
      required: true,
    })
  };

  async run() {
    const { flags } = this.parse(GetFeed);
    await handler({ command: this, query: GetFeedDocument, variables: flags });
  }
}
"
`;

exports[`oclif cli builds from a read query with single enum parameter 1`] = `
"
const GetFeedDocument = \`
query GetFeed($type: FeedType!) {
  feed(type: $type) {
    id
    commentCount
  }
}\`

export default class GetFeed extends Command {
  
  
  static flags = {
    help: flags.help({ char: 'h' }),
    type: flags.string({
      multiple: false,
      required: true,
    })
  };

  async run() {
    const { flags } = this.parse(GetFeed);
    await handler({ command: this, query: GetFeedDocument, variables: flags });
  }
}
"
`;

exports[`oclif cli builds from a read query with single string parameter 1`] = `
"
const GetEntryDocument = \`
query GetEntry($name: String!) {
  entry(repoFullName: $name) {
    repository {
      name
    }
  }
}\`

export default class GetEntry extends Command {
  
  
  static flags = {
    help: flags.help({ char: 'h' }),
    name: flags.string({
      multiple: false,
      required: true,
    })
  };

  async run() {
    const { flags } = this.parse(GetEntry);
    await handler({ command: this, query: GetEntryDocument, variables: flags });
  }
}
"
`;

exports[`oclif cli reads metadata from the @oclif directive 1`] = `
"
const GetCurrentUserDocument = \`
query GetCurrentUser {
  currentUser {
    login
  }
}\`

export default class GetCurrentUser extends Command {
  
static description = \\"Get Current User\\";

  
static examples: string[] = [\\"cli get-current-user\\",\\"hello\\"];

  static flags = {
    help: flags.help({ char: 'h' }),
    
  };

  async run() {
    const { flags } = this.parse(GetCurrentUser);
    await handler({ command: this, query: GetCurrentUserDocument, variables: flags });
  }
}
"
`;
